community/kernel/src/main/java/org/neo4j/kernel/impl/locking/deferred/DeferringLockClient.java

```diff
@@ -0,0 +1,198 @@
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.locking.deferred;
+
+import org.apache.commons.lang3.mutable.MutableInt;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.TreeMap;
+
+import org.neo4j.kernel.impl.locking.AcquireLockTimeoutException;
+import org.neo4j.kernel.impl.locking.LockClientStoppedException;
+import org.neo4j.kernel.impl.locking.Locks;
+
+// TODO the state keeping in this class is quite unoptimized, please do so
+public class DeferringLockClient implements Locks.Client
+{
+    private final Locks.Client clientDelegate;
+    private final Map<LockUnit,MutableInt> locks = new TreeMap<>();
+    private volatile boolean stopped;
+
+    public DeferringLockClient( Locks.Client clientDelegate )
+    {
+        this.clientDelegate = clientDelegate;
+    }
+
+    @Override
+    public void acquireShared( Locks.ResourceType resourceType, long... resourceIds ) throws AcquireLockTimeoutException
+    {
+        for ( long resourceId : resourceIds )
+        {
+            addLock( resourceType, resourceId, false );
+        }
+    }
+
+    @Override
+    public void acquireExclusive( Locks.ResourceType resourceType, long... resourceIds )
+            throws AcquireLockTimeoutException
+    {
+        for ( long resourceId : resourceIds )
+        {
+            addLock( resourceType, resourceId, true );
+        }
+    }
+
+    @Override
+    public boolean tryExclusiveLock( Locks.ResourceType resourceType, long resourceId )
+    {
+        throw new UnsupportedOperationException( "Should not be needed" );
+    }
+
+    @Override
+    public boolean trySharedLock( Locks.ResourceType resourceType, long resourceId )
+    {
+        throw new UnsupportedOperationException( "Should not be needed" );
+    }
+
+    @Override
+    public void releaseShared( Locks.ResourceType resourceType, long resourceId )
+    {
+        removeLock( resourceType, resourceId, false );
+    }
+
+    @Override
+    public void releaseExclusive( Locks.ResourceType resourceType, long resourceId )
+    {
+        removeLock( resourceType, resourceId, true );
+    }
+
+    void acquireDeferredLocks()
+    {
+        long[] current = new long[10];
+        int cursor = 0;
+        Locks.ResourceType currentType = null;
+        boolean currentExclusive = false;
+        for ( LockUnit lockUnit : locks.keySet() )
```

Comment:
Splitting the locks per type would make release and acquire more complex and we do not see how splitting will allow a more compact representation because we would still need to have some map structure to make release work efficient.